File: include/blocks/Cpu.c
#include <blocks.h>
#include <stdio.h>
#include <utils.h>

static long double pcpu[7] = {0, 0, 0, 0, 0, 0, 0};

int blk_cpu(char *value) {
  long long cpu[7], used, total, prev_total = 0, new_total = 0;

  if (fpscanf("/proc/stat", "%*s %lld %lld %lld %lld %lld %lld %lld", &cpu[0],
              &cpu[1], &cpu[2], &cpu[3], &cpu[4], &cpu[5], &cpu[6]) != 7)
    return -1;

  for (int i = 0; i < 7; ++i)
    new_total += cpu[i];
  for (int i = 0; i < 7; ++i)
    prev_total += pcpu[i];

  total = new_total - prev_total;
  used = cpu[0] + cpu[1] + cpu[2] - pcpu[0] - pcpu[1] - pcpu[2];

  for (int i = 0; i < 7; ++i)
    pcpu[i] = cpu[i];

  int percent = (100 * used) / total;
  return percent <= 25 ? sprintf(value, "  %2d%%", percent)
         : percent > 25 && percent <= 65
             ? sprintf(value, "  <Fg=" YELLOW ">%2d</Fg>%%", percent)
             : sprintf(value, "  <Fg=" RED ">%2d</Fg>%%", percent);
}

File: src/loggers/cpu.rs
//! Logs cpu usage.
use super::{Logger, ValueRunner};
use std::{
    convert::TryInto,
    fs::File,
    io::{BufRead, BufReader},
};

struct CpuRunner {
    previous: [u64; 7],
}

impl CpuRunner {
    fn fmt_value(string: String) -> String {
        format!(
            "<BtnL=notify_max_cpu> {}  </BtnL><Box:Left=#171717:2> </Box>",
            string
        )
    }

    fn calculate(&mut self) -> Option<String> {
        let file = File::open("/proc/stat").ok()?;
        let mut line = String::new();
        BufReader::new(file).read_line(&mut line).ok()?;

        let cpu: Vec<u64> = line
            .split_whitespace()
            .skip(1)
            .take(7)
            .map(|n| n.parse().unwrap())
            .collect();

        let total = {
            let new_total: u64 = cpu.iter().sum();
            let prev_total: u64 = self.previous.iter().sum();
            new_total - prev_total
        };
        let used = {
            let new_used: u64 = cpu.iter().take(3).sum();
            let prev_used: u64 = self.previous.iter().take(3).sum();
            new_used - prev_used
        };
        self.previous = cpu.try_into().unwrap();

        Some((100 * used) / total)
            .and_then(|percent| match percent {
                0..=25 => Some(format!("  {:3}%", percent)),
                26..=65 => Some(format!("  <Fg=#ffdd59>{:3}</Fg>%", percent)),
                66..=100 => Some(format!("  <Fg=#cc6666>{:3}</Fg>%", percent)),
                _ => None,
            })
            .map(Self::fmt_value)
    }
}

impl ValueRunner for CpuRunner {
    fn get_value(&mut self) -> String {
        self.calculate()
            .or_else(|| Some(CpuRunner::fmt_value("cpu: ?".into())))
            .unwrap()
    }
}

pub fn create_cpu_logger() -> Logger {
    Logger::ValueLogger {
        interval_ms: 1000,
        runner: Box::new(CpuRunner { previous: [0; 7] }),
    }
}
